### ISSUE TITLE
Media privacy

### ISSUE THREAD

**Issue by Gawrion** [HuubMons/HuMo-genealogy#47]
Hey. I think that You should implement some mechanism to protect media (photos) from access from guest accounts. Now i can grab url of photo simply pasting its filename from media directory.  It can be done by changing directory of media outside www root and make php to grab it dependant on account rights. There are many possiblities but i think that You should adress it.

**Comment by David-Baron** [comment]
Hello.
On a web application, you can not protect the media, they are necessarily public with a url.
Currently you have a possibility:
- make a public tree without the photos
- make a tree with photos and an identification for the people to whom you want to give access.
Another solution would be to have only one tree with an identification option to see the photos, I do not think it exists and I know that the author of the application is already overloaded by other modifications.

This not being a "problem" I invite you to go to the [forum](https://humo-gen.com/genforum/) where you will find more help and advice for using this application.

**Comment by Gawrion** [comment]
Hey David. Of course u can protect them. For example: u place photos outside a www root directory and serve them dynamically by php script that checks if u re authorized to use. It's not just only theory - i use protecting files this way at my website. I can provide some example code.

**Comment by David-Baron** [comment]
So re read what I told you. Your method is not applicable to humo-genealogy because many users use SEO, which your method does not respect, so I give you the currently applicable solutions.

**Comment by Gawrion** [comment]
Hey. Protecting photos by having a tree without photos is absurdal solution... Sorry but none of your solutions solves my problem and none adresses the issue proffesionally. 
Other thing is that u don't have time to implement, don't want to do that - i'm not urging to do that. I just adressing an important issue according to GDPR law and privacy. Privacy issueas should always have high priority. And other thing is that there is no problem for coding that. 
Of course serving private protected files also involves making a custom code to save those files in a certain directory, but it's as EZ as serving them.

Here is some code in separate php file i use to serve png/jpg files based on URI parameter for "file name":
```
<?php
session_start();
if (isset($_SESSION['loggedin']) && isset($_GET['picture']) && isset($_GET['uid']) && $_GET['uid'] === $_SESSION['uid']) {
    $directory= 'directory_with_all_private_files';
    //users directory
    $dir = $directory. "/" . $_SESSION['uid'];
    $dir = realpath($dir);
    $picture= $dir . '/' . basename($_GET['picture']);
    $picture= realpath($picture);
    if (is_file($picture)) {
        header('Content-Type: image/png');
        header('Cache-Control: private, max-age=3600');
        header('Pragma:');
        header('Expires: ' . gmdate('D, d M Y H:i:s \G\M\T', time() + (60 * 60))); // 1 hour cache
        readfile($picture);
    }
    } else {
    //global function to give 404 header
    ustawHeader404();
    }
```

Implementing that function/way should be as EZ as parsing file name (u have that already in database under humo_events (event_event) by above function/php file. Maybe some minor changes should be done also.

When I want to give URL to a file i simply use it that way. This is a real file URL i get when im logged in. U will get 404 ("obrazek" means "picture" - i changed variables names to english so u can ez understand the code):
https://my.oodo.pl/modules/profile/daj_obrazek.php?obrazek=9b4a10e70eaf3b951eaa542410a45040.png&uid=37

**Comment by David-Baron** [comment]
I think you didn't understand my message but it's not important.
On the other hand:
"I just addressing an important issue according to GDPR law and privacy" has nothing to do with this unless you use and distribute images that you are not the author of or that you do not have the authorizations.
"global function to give 404 header"
This does not correspond to a clean code, should be "401 Unauthorized"
In addition, the majority of users index their site and their images, therefore contrary to the current use of humo-gen.
Nothing prevents you from doing a search to find an application that corresponds to your needs, or possibly doing it yourself.

**Comment by Gawrion** [comment]
1. Im lawyer specialized in GDPR topics. U (or other authorized ppl) can have legal basis for personal use of some photos under article 2 GDPR, but giving access to others will be violation of GDPR. It's obvious for me. In similar way u share some private rosources with ppl u want but not whole world. Obvious things.
2. Give header as U wish. Good practice is to inform not authorized user that resource do not exists than that it exists and he can find a workaround way to get it. But give headers as u wish. I use 404. For example for password retrieval i also do not inform that e-mail is/insn't in database ;) Do as u wish.

HumoGen is an excelent tool - i dont need other. I can contribute to this tool and make it better. U seem to be kind of person that rather finds a problems, not solutions. Not a good person to represent with a high standards. U re rather rude than helpfull.

**Comment by David-Baron** [comment]
1 - So you are (maybe) a lawyer, I invite you to read and reread the scope of this article 2 paragraph 2c and you will then understand that it does not apply. In addition, humogen has a function that corresponds to the GDPR to hide the personal data of the people you want.
2 - For SEO, you must indicate why some images are available and others not to the browser.
3 - "Privacy standards are evolving - you are talking about standards from 2 decades ago" you are talking about something that you do not yet master because I saw a uid pass in GET in your code.

In fact, I just came to give you "palliative" solutions that you interpret in an incomprehensible way, sorry but I can't do anything for you.

**Comment by HuubMons** [comment]
At this moment i'm having a small vacation. I will check this thread and
added code later.

Op zo 17 nov 2024 11:34 schreef David Baron ***@***.***>:

**Comment by Gawrion** [comment]
David - sorry but i don't want to talk with u more. I do not see the point. 

HuubMoons - thanks. I can do some code changes and tests, but for now i don't know the philosophy of code. Before i'll do that it would be nice to hear few words about the code logic from U, rather than analyze it by myself.

**Comment by David-Baron** [comment]
Your code example is an old code that is more than 15 years old and is not secure, so we will not put it in humogen. So I reassure you, you have nothing to do, the code is almost finished, it will soon be available to Huub. 
But thank you for your petite participation.

**Comment by Gawrion** [comment]
Lol. David. U can't put that code to humogen - it's only example of technique u can serve the files. In Poland we call ppl like You: "głupiomądry".

**Comment by Gawrion** [comment]
One more thing. If you're serving file dynamically by php with headers i gave in above example, when u try to save the file (rendered inline in browser) it will have name of php file. Now i think in humogen it will be more usefull when file would have original name when saved. U can do it adding:
`header('Content-Disposition: inline; filename="' . $filename . '"');`

**Comment by HuubMons** [comment]
> HuubMons - thanks. I can do some code changes and tests, but for now i don't know the philosophy of code. Before i'll do that it would be nice to hear few words about the code logic from U, rather than analyze it by myself.

I'm rebuilding all scripts. The original scripts were mixed PHP and HTML (the first version was around 2005, there wasn't CSS and seperated scripts then. We used less scripts because it was faster reading these files at webservers).

Several scripts are now extracted into MVC model (under construction):
index.php -> link to controller.
controller script -> get data from model script.
index.php -> shows view script.

At this moment I'm not convinced about moving photos outside the www root directory.

1) If a user sees a photo it's always possible to copy this photo. Because this photo is allready in the cache of the browser, and it's possible to just create a screenshot.

2) There is allready a privacy filter build in HuMo-genealogy. If the privacy filter is active, photo's will not be shown.

3) Photo's are shown using glightbox (maybe replaced with bootstrap in the future). I'm not sure if an added script to secure photo's will work in combination with glightbox.

4) All photo scripts will be changed. Because of this forum topic: https://humo-gen.com/genforum/viewtopic.php?t=3069

But: I actually don't care to add a new function to HuMo-genealogy if needed for security.

**Comment by Gawrion** [comment]
Finally some words i fully understand ;)  

I saw this controllers logic. Seems straight and clean, but i will need time to analyze it.

Of course u know all the relations and potential problems better than me. Now - when i work with a new version i see that its some hybrid (old, static html and some new things in ajax). Not whole UX is nice - that confirmation at the top is most annoying.

1. Dynamically serving photos behaves (as far as i know) the same as static - u can also set header for cache. U can always save the photo. Difference is only URL with php parameter. What screenshot  do U mean?
2. Privacy filter do not protect photos. It only prevents from displaying. U can still download a photo knowing the name of file. Im afraid of some brute force crawlers etc.
3. It's been a long time i touched lightbox (last time few years ago in Drupal). Now i checked - asked GPT if Lightbox2 can create it's show with dynamic links - GPT - says it can - it only needs valid url, no matter if it;s static or dynamic.
<img src="https://github.com/user-attachments/assets/a771df1d-f6b6-4e63-aa20-a3d13945b48f" sidth="50">
4. OK. I see u have some work to do with thumbnails. I agree that logic where: photo uploaded=thumbnail generated - is much more better, and there wont be timeouts with large media dir. Agree. But for thumnails u can apply the same logic as for the original files - the same parser function checking the same privileges (logically its the same media, but smaller ;))

Ok. I understand that u don;t want to adress that issue, but im, a little bit concerned. In a few weeks i'll try to analyze the code and do some stuff and maybe create pull request after testing or fork. 

When U plan to do some bigger changes to code, so we can do some synergy?

**Comment by HuubMons** [comment]
> When U plan to do some bigger changes to code, so we can do some synergy?

It's a hobby project, I don't have a specific planning. Most recent development version can be found here:
https://github.com/HuubMons/HuMo-genealogy/tree/prepare_next_version

**Comment by Gawrion** [comment]
Sure! Take your time ;)

BTW. With a new version i got some warnings and fatal errors on admin thumbnails page. I dont know if it's related problem but random photos on the main page are not showing new added images.

Warning: Undefined variable $data2Db in /var/www/gene-tots/admin/views/thumbs.php on line 500

Warning: Attempt to read property "tree_pict_path" on null in /var/www/gene-tots/admin/views/thumbs.php on line 500

Deprecated: substr(): Passing null to parameter #1 ($string) of type string is deprecated in /var/www/gene-tots/admin/views/thumbs.php on line 501

Fatal error: Uncaught ValueError: imagecreatetruecolor(): Argument #1 ($width) must be greater than 0 in /var/www/gene-tots/admin/views/thumbs.php:565 Stack trace: #0 /var/www/gene-tots/admin/views/thumbs.php(565): imagecreatetruecolor() #1 /var/www/gene-tots/admin/index.php(747): include_once('...') #2 {main} thrown in /var/www/gene-tots/admin/views/thumbs.php on line 565

**Comment by HuubMons** [comment]
The thumbs scripts are being rebuild: https://humo-gen.com/genforum/viewtopic.php?t=3069

Nice topola viewer.

**Comment by Gawrion** [comment]
I fixed the problem replacing:
$imtype = strtoupper(substr($file, -3));

by:
if (mime_content_type($folder . $file) == 'image/jpeg' OR mime_content_type($folder . $file) == 'image/pjpeg') {
    $imtype = 'JPG';
    } elseif (mime_content_type($folder . $file) == 'image/png') {
    $imtype = 'PNG';
    } elseif (mime_content_type($folder . $file) == 'image/gif') {
    $imtype = 'GIF';
    }

It works but it is not imho optimized as i can see code has already function (check_media_type) that checks mime but returns boolean. It could be refactorized and optimized and maybe changed to return false or mime type and serve in this other function to determine image type.

**Comment by Gawrion** [comment]
I know why code doesn't show my new photos in random_photo() function. There is string replacement spaces to underscores. IDK why. It takes filename from evennt_event and changes filename so in while loop it can't find file by file_exists function. I changed:
$picname = str_replace(" ", "_", $picqryDb->event_event);
to:
$picname = $picqryDb->event_event;
and finally i got my all photos in random photo views ;)

**Comment by Gawrion** [comment]
I also added possibility to got the link from picture with no date and no text by replacing:
$text .= '<a href="' . $url . '">' . $date_place . $picqryDb->event_text . '</a></div>';
by:
if ($picqryDb->event_text !== '' OR $date_place !== '' ) {
   $text .= '<a href="' . $url . '">' . $date_place . $picqryDb->event_text . '</a></div>';
} else {
   $text .= '<a href="' . $url . '">Go to person&apos;s page</a></div>';
}
Now i can go to person's page from every photo. U would need to make transaltions for this line working - i don't know how this language system works.

**Comment by HuubMons** [comment]
> I know why code doesn't show my new photos in random_photo() function. There is string replacement spaces to underscores. IDK why. It takes filename from evennt_event and changes filename so in while loop it can't find file by file_exists function. I changed: `$picname = str_replace(" ", "_", $picqryDb->event_event);` to: `$picname = $picqryDb->event_event;` and finally i got my all photos in random photo views ;)
> 
> Maybe u remember what for was that space to underscore replacement? Obviously there was some reason :/ Maybe that is enviorment/system dependant?

This is old code, but it is done to avoid a space in the url to the picture. And that doesn't work at internet.
Just never use spaces in file names or picture names at a website.

**Comment by Gawrion** [comment]
Hey Huub! Ok - now I understand the reason.

Files with spaces in names work good. For main images and for thumbs generated with new code too:
https://gene.tots.pl/media/Helena%20Biarda%20akt%20urodzenia%202.jpg
https://gene.tots.pl/media/thumb_Helena%20Biarda%20akt%20urodzenia%202.jpg.jpg

I agree that underscores are better but at the other hand end user will not care as even i don't care. If u want to ensure that files doesn't have spaces it would be better to validate it while upploading and then replace spaces by underscores. Whe can achieve that easily by putting str_replace in uploading part.

Now script allows to write files with spaces. It saves names with spaces as files and in database, but then it takes name from database and replaces spaces with underscores which leads to "no file" when checking if file exists.

**Comment by HuubMons** [comment]
> I also added possibility to got the link from picture with no date and no text by replacing: `$text .= '<a href="' . $url . '">' . $date_place . $picqryDb->event_text . '</a></div>';` by:
> 
> ```
> if ($picqryDb->event_text !== '' OR $date_place !== '' ) {
>    $text .= '<a href="' . $url . '">' . $date_place . $picqryDb->event_text . '</a></div>';
> } else {
>    $text .= '<a href="' . $url . '">Go to person&apos;s page</a></div>';
> }
> ```
> 
> Now i can go to person's page from every photo. U would need to make transaltions for this line working - i don't know how this language system works.


First: thank you for testing the code! I really appreciate it!

I'm a bit puzzled about this. Normally date/ place or name of person is used in the picture link. Is there really no date, place or name of the person?

But I don't mind to add this code, I will test this and add it to the code.

**Comment by Gawrion** [comment]
Yeah - in random photo function it didn't show any text when there were no date AND no text. So i added this url text. It would be better to take name And surname of person and create 'Go to $name_and_surname page' - but i didn't know how to get it. I know that is some property in persons class but u will know i t better ;)
I like to test it becouse i use it last weeks. Today i discovered some new documents about my Grandmother. I really like to reveal the past about my ancestors!

**Comment by HuubMons** [comment]
Yes, name is probably better! I don't know why it's missing, I will check this!

**Comment by HuubMons** [comment]
Done! Just added the code to Github.

**Comment by Gawrion** [comment]
I made quick POC for media privacy feature:
https://humo-gen.com/genforum/viewtopic.php?t=3085

**Comment by Gawrion** [comment]
Topic has update: https://humo-gen.com/genforum/viewtopic.php?p=17744#p17744

**Comment by Gawrion** [comment]
Im working today on functions checking if U can safely enable "media privacy mode". Work in progress ;)
![Screenshot_36](https://github.com/user-attachments/assets/77790183-fb74-4866-85c9-c6e6957e7493)

**Comment by Gawrion** [comment]
Update:
![Uploading Screenshot_37.png…]()

**Comment by HuubMons** [comment]
Nice!

Be careful it isn't too complex. If generating an htaccess file fails, it's possible to just show this text for example:

Automatic creation of htaccess file failed. You could create the htaccess file, and upload to the media directory. The htaccess file should contain this code:

$htaccessContent = <<<HTACCESS
        <FilesMatch ".*">
          #Require all denied
        </FilesMatch>
        HTACCESS;

**Comment by Gawrion** [comment]
Ok. I made PR - u can check the code and change it if needed.

**Comment by HuubMons** [comment]
I have merged the changes into a seperate new branch at Github "secure_media_folder".

When I try to enable the media security, I have this error message:
`I can't find ../humo-gen-afb/mons/ directory. Check if it's created.`

I personally always use a media path outside the HuMo-genealogy folder. And I always recommend users to do the same.
At this moment I haven't check the code.

**Comment by HuubMons** [comment]
Using docker at Windows laptop:

Scripts in:
http://127.0.0.1/humo-genealogy/
absolute path: /var/www/html/humo-genealogy/

Media in:
http://127.0.0.1/humo-gen-afb/mons/
absolute path: /var/www/html/humo-gen-afb/mons/

Document root:
/var/www/html/

To prevent problems using Windows, Linux etc. I'm always using relative paths.

I always recommend to use a subfolder if HuMo-genealogy is used. Otherwise other programs like Wordpress, Joomla etc. could mess up the program (it happened before).
The HuMo-genealogy website is also using a subfolder: https://humo-gen.com/humo-gen/

**Comment by Gawrion** [comment]
Thx Huub.. As i suspected. Fixed code updated. Now it should work in dir relative upper than app files but still in root and completely outside root for even better security. I tested it for such config.

PR: https://github.com/HuubMons/HuMo-genealogy/pull/66

**Comment by HuubMons** [comment]
I wanted to merge the media_security branch into the prepare_next_version branch.
It seems I used the wrong command, I lost your changed file. Could you submit the changes to the prepare_next_version branch?

**Comment by Gawrion** [comment]
Ok. I will try to do that.

**Comment by HuubMons** [comment]
Something like that happened to me. It there is nothing to compare, but I'm quit sure you changed a file somewhere.

**Comment by Gawrion** [comment]
Which file/changes You mean? If u need any help/info/me to do sthng - just write.

**Comment by HuubMons** [comment]
If a thumb is missing (I'm testing a ancestor report), then page is slow.
There is an error message: 
GET http://127.0.0.1/images/thumb_missing-image.jpg 404 (Not Found)

1) Part of the path is missing.
2) The original media file should be shown (in same size as a thumbnail). So there will be always a media file shown even if there are no thumbnails.
